### PR TITLE
MCR-3115 allow parallel indexing in different cores

### PR DESCRIPTION
**Pull request by sebhofmann**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3115).

based on https://github.com/MyCoRe-Org/mycore/pull/2167

change base to main if https://github.com/MyCoRe-Org/mycore/pull/2167 is merged

### CODE REVIEW

File: mycore-solr/src/main/java/org/mycore/solr/MCRSolrCore.java [hunk 1]
```diff
@@ -110,18 +102,20 @@ public MCRSolrCore(String serverURL, String name) {
      * @param shardCount
      *            number of shards
      */
-    MCRSolrCore(String serverURL, String name, String configSet, Integer shardCount) {
-        setup(serverURL, name, configSet, shardCount);
+    public MCRSolrCore(String serverURL, String name, String configSet, Integer shardCount, Set<MCRSolrCoreType> type) {
```
Comment (toKrause) [inline mycore-solr/src/main/java/org/mycore/solr/MCRSolrCore.java]: Last parameter should be names `types`

File: mycore-solr/src/main/java/org/mycore/solr/MCRSolrCore.java [hunk 2]
```diff
@@ -110,18 +102,20 @@ public MCRSolrCore(String serverURL, String name) {
...
+        setup(serverURL, name, configSet, shardCount, type);
     }
 
-    protected void setup(String serverURL, String name, String configSet, Integer shardCount) {
+    protected void setup(String serverURL, String name, String configSet,
```
Comment (toKrause) [inline mycore-solr/src/main/java/org/mycore/solr/MCRSolrCore.java:110]: Last parameter should be names `types`

File: mycore-solr/src/main/java/org/mycore/solr/index/MCRSolrIndexEventHandler.java [hunk 3]
```diff
@@ -187,12 +190,12 @@ protected void handlePathDeleted(MCREvent evt, Path file, BasicFileAttributes at
     protected void updateDerivateFileIndex(MCREvent evt, MCRDerivate derivate) {
         MCRSessionMgr.getCurrentSession().onCommit(() -> {
             //MCR-2349 initialize solr client early enough
-            final SolrClient mainSolrClient = MCRSolrClientFactory.getMainSolrClient();
+            final List<MCRSolrCore> mainSolrClient = MCRSolrCoreManager.getCoresForType(MCRSolrCoreType.MAIN);
```
Comment (toKrause) [inline mycore-solr/src/main/java/org/mycore/solr/index/MCRSolrIndexEventHandler.java]: `mainSolrClient` should be renamed to, maybe `mainSolrCores`

File: mycore-solr/src/main/java/org/mycore/solr/index/MCRSolrIndexer.java [hunk 4]
```diff
@@ -324,8 +326,8 @@ protected static boolean useNestedDocuments() {
     /**
      * Rebuilds solr's metadata index.
      */
-    public static void rebuildMetadataIndex(SolrClient solrClient) {
-        rebuildMetadataIndex(MCRXMLMetadataManager.instance().listIDs(), solrClient);
+    public static void rebuildMetadataIndex(List<MCRSolrCore> solrCore) {
```
Comment (toKrause) [inline mycore-solr/src/main/java/org/mycore/solr/index/MCRSolrIndexer.java]: Should be `solrCores`

File: mycore-solr/src/main/java/org/mycore/solr/index/handlers/stream/MCRSolrFilesIndexHandler.java [hunk 5]
```diff
@@ -108,8 +107,8 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
         });
         int fileCount = subHandlerList.size() + docs.size();
         LOGGER.info("Sending {} file(s) for derivate \"{}\"", fileCount, derivateID);
-        if (!docs.isEmpty()) {
-            MCRSolrInputDocumentsHandler subHandler = new MCRSolrInputDocumentsHandler(docs, solrClient);
+        if (!docs.isEmpty()) { // TODO: check this crap
```
Comment (toKrause) [inline mycore-solr/src/main/java/org/mycore/solr/index/handlers/stream/MCRSolrFilesIndexHandler.java]: Could use a better description of what exactly needs to be done here

**Review by toKrause — CHANGES_REQUESTED** [review]
Only some minor stuff.

**Review by toKrause — DISMISSED** [review]

**Review by rsteph-de — APPROVED** [review]
looks good for me

**Pull request merged**